data-structures/filecoin/schema.md

```diff
@@ -0,0 +1,989 @@
+  map Bytes
+  data [ DealOpsByEpochHAMTSetElement ]
+} representation tuple
+
+type DealOpsByEpochHAMTSetElement union {
+  | DealOpsByEpochHAMTSet_Link "0"
+  | DealOpsByEpochHAMTSetBucket "1"
+} representation keyed
+
+type DealOpsByEpochHAMTSet_Link &DealOpsByEpochHAMTSet
+
+type DealOpsByEpochHAMTSetBucket [ DealOpsByEpochHAMTSetBucketEntry ]
+
+type DealOpsByEpochHAMTSetBucketEntry struct {
+  key Bytes # The bytes of the string form of ChainEpoch
+  value Null
+} representation tuple
+```
+
+### StorageMarketActor
+
+```ipldsch
+type MinerV0State struct {
+  Info  &MinerV0Info
+  PreCommitDeposits TokenAmount
+  LockedFunds TokenAmount
+  VestingFunds &MinerV0VestingFunds
+  InitialPledge TokenAmount
+  PreCommittedSectors &MinerV0SectorPreCommitOnChainInfoHAMT # HAMT[SectorNumber]SectorPreCommitOnChainInfo
+  PreCommittedSectorsExpiry &BitFieldQueueAMT # AMT[ChainEpoch]BitField
+  AllocatedSectors &BitField
+  Sectors &MinerV0SectorOnChainInfoAMT # AMT[SectorNumber]SectorOnChainInfo
+  ProvingPeriodStart ChainEpoch
+  CurrentDeadline Int
+  Deadlines &MinerV0Deadlines
+  EarlyTerminations BitField
+} representation tuple
+```
+
+**v2**
+
+```ipldsch
+type MinerV2State struct {
+  Info  &MinerV2Info
+  PreCommitDeposits TokenAmount
+  LockedFunds TokenAmount
+  VestingFunds &MinerV0VestingFunds
+  FeeDebt TokenAmount
+  InitialPledge TokenAmount
+  PreCommittedSectors &MinerV0SectorPreCommitOnChainInfoHAMT # HAMT[SectorNumber]SectorPreCommitOnChainInfo
+  PreCommittedSectorsExpiry &BitFieldQueueAMT # AMT[ChainEpoch]BitField
+  AllocatedSectors &BitField
+  Sectors &MinerV0SectorOnChainInfoAMT # AMT[SectorNumber]SectorOnChainInfo
+  ProvingPeriodStart ChainEpoch
+  CurrentDeadline Int
+  Deadlines &MinerV2Deadlines
+  EarlyTerminations BitField
+} representation tuple
+```
+
+```ipldsch
+type MinerV0Info struct {
+  Owner Address
+  Worker Address
+  ControlAddresses nullable [Address]
+  PendingWorkerKey nullable MinerV0WorkerChangeKey
+  PeerId PeerID
+  Multiaddrs nullable [Multiaddr]
+  SealProofType Int
+  SectorSize SectorSize
+  WindowPoStPartitionSectors Int
+} representation tuple
+
+type MinerV0WorkerChangeKey struct {
+  NewWorker Address
+  EffectiveAt ChainEpoch
+} representation tuple
+```
+
+**v2**
+
+```ipldsch
+type MinerV2Info struct {
+  Owner Address
+  Worker Address
+  ControlAddresses nullable [Address]
+  PendingWorkerKey nullable MinerV0WorkerChangeKey
+  PeerId PeerID
+  Multiaddrs nullable [Multiaddr]
+  SealProofType Int
+  SectorSize SectorSize
+  WindowPoStPartitionSectors Int
+  ConsensusFaultElapsed ChainEpoch
+  PendingOwnerAddress nullable Address
+} representation tuple
+```
+
+```ipldsch
+type MinerV0VestingFunds struct {
+  Funds [MinerV0VestingFund]
+} representation tuple
+
+type MinerV0VestingFund struct {
+  Epoch ChainEpoch
+  Amount TokenAmount
+} representation tuple
+```
+
+```ipldsch
+type MinerV0Deadlines struct {
+  Due MinerV0DeadlineLinkList
+} representation tuple
+
+# Must be 48 CIDs
```

Comment:
ok, so this might be going a bit too far but I found this useful when validating block forms, I could just wind this back to `type MinerV0DeadlineLinkList [&MinerV0Deadline] # must be exactly 48 entries`.

Reply:
- that one did kind of scare me as I scrolled through this at high speed
- it's also a pretty interesting example where indeed either of these two ways to write it is 100% valid.  Both ways are a schema that matches the data!